torch/_dynamo/variables/builder.py

```diff
@@ -1284,8 +1284,8 @@ def _clone_input(value):
 
         # Handle recursive calls here
         elif (
-            isinstance(example_value, FakeTensor)
-            and example_value.fake_mode is tx.fake_mode
+            is_fake(example_value)
+            and maybe_get_fake_mode(example_value) is tx.fake_mode
```

Comment:
I think I'm following the design of https://github.com/pytorch/pytorch/pull/105308 where a tensor subclass is not fakefied, but its children are.  If this is true, we need to check for fake mode on the children, which is what the `maybe_get_fake_mode` function is doing.

I do agree this is a bit confusing.  For now I'm just trying to patch some code which doesn't seem to follow the design of https://github.com/pytorch/pytorch/pull/105308, I don't fully understand why we don't fakify the parent class.  There probably is a good reason.

Reply:
Hmm yeah @voznesenskym tell me if anything here sounds unreasonable, but this is basically what you have to do now to handled "fakeified torch_dispatch subclasses".

```
subclass_tensor: MySubclass(torch.ones(2, 2))
fakeified_subclass_tensor: MySubclass(FakeTensor(2, 2))
```

When we fakeifiy a "wrapper tensor sublcass", the result is not a fake tensor - it's another wrapper tensor subclass, that now holds fake tensors instead of real tensors. And yep this can apply recursively if you have a wrapper subclass holding another wrapper subclass - `is_fake` and `maybe_get_fake_mode` take care of that.

So one consequence of the above is that you can no longer blindly call `fake_tensor_mode.from_tensor(my_subclass).fake_mode` (you need to "find" the mode nested on the leaf fake tensor somewhere inside)

Reply:
I think the reason we chose to fakify the inner tensor is that even if we fakify the wrapper tensor, the inner dense tensor also need to be fakified in order to avoid materializing memory. Also there's quite some complexity to get wrapper subclass fakification to go through existing fake tensor propagation (i.e. map storaging alias, etc.)

Reply:
Yep - we also **want** the fakeified tensors to be subclasses - so when AOTAutograd gets them, it sees some actual tensor subclasses that it can trace through the __torch_dispatch__ of.